sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -3864,31 +3890,62 @@ pi_result cuda_piEnqueueEventsWaitWithBarrier(pi_queue command_queue,
     return PI_ERROR_INVALID_QUEUE;
   }
 
+  pi_result result;
+
   try {
     ScopedContext active(command_queue->get_context());
+    pi_uint32 stream_token;
+    _pi_stream_guard guard;
+    CUstream cuStream = command_queue->get_next_compute_stream(
+        num_events_in_wait_list, event_wait_list, guard, &stream_token);
+    {
+      std::lock_guard(command_queue->barrier_mutex_);
+      if (command_queue->barrier_event_ == nullptr) {
+        PI_CHECK_ERROR(cuEventCreate(&command_queue->barrier_event_,
+                                     CU_EVENT_DISABLE_TIMING));
+      }
+      if (num_events_in_wait_list == 0) { //  wait on all work
+        if (command_queue->barrier_tmp_event_ == nullptr) {
+          PI_CHECK_ERROR(cuEventCreate(&command_queue->barrier_tmp_event_,
+                                       CU_EVENT_DISABLE_TIMING));
+        }
+        command_queue->sync_streams(
+            [cuStream,
+             tmp_event = command_queue->barrier_tmp_event_](CUstream s) {
+              if (cuStream != s) {
+                PI_CHECK_ERROR(cuEventRecord(tmp_event, s));
+                PI_CHECK_ERROR(cuStreamWaitEvent(cuStream, tmp_event, 0));
```

Comment:
Very clever! Could you please comment what is happening in the new changes in this function?

Reply:
It makes one stream wait on all other streams (creating a new cuEvent for every stream and recording it in that stream before waiting on it in the chosen stream). Future work will, when enqueued to a stream for the first time, wait on the stream that waited on all other work.

Reply:
I meant add comments in the code to make it easier to understand for other maintainers when they pass by this. 😄